File: QytetetJava/src/modeloqytetet/Qytetet.java
package modeloqytetet;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 *
 * @author aurelia
 */
public class Qytetet {
    private static final Qytetet instance = new Qytetet();
    public static final int MAX_JUGADORES = 4;
    static final int MAX_CARTAS = 10;
    static final int MAX_CASILLAS = 20;
    static final int PRECIO_LIBERTAD = 200;
    static final int SALDO_SALIDA = 1000;
    private Sorpresa cartaActual;
    private ArrayList<Sorpresa> mazo;
    private ArrayList<Jugador> jugadores;
    private Jugador jugadorActual;
    private Tablero tablero;
    private Dado dado;
    
    private Qytetet(){
        cartaActual = null;
        mazo = new ArrayList();
        jugadores = new ArrayList();
        jugadorActual = null;
        tablero = null;
        dado = Dado.getInstance();
    }
    
    public static Qytetet getInstance(){
        return instance;
    }
    
    public boolean aplicarSorpresa(){
        boolean tienePropietario = false;
        if (cartaActual.getTipo() == TipoSorpresa.PAGARCOBRAR){
            jugadorActual.modificarSaldo(cartaActual.getValor());
        }else if(cartaActual.getTipo() == TipoSorpresa.IRACASILLA){
            boolean esCarcel = tablero.esCasillaCarcel(cartaActual.getValor());
            if (esCarcel){
                encarcelarJugador();
            }else{
                Casilla nuevaCasilla = tablero.obtenerCasillaNumero(cartaActual.getValor());
                tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
            }
        }else if (cartaActual.getTipo() == TipoSorpresa.PORCASAHOTEL){
            jugadorActual.pagarCobrarPorCasaYHotel(cartaActual.getValor());
        }else if (cartaActual.getTipo() == TipoSorpresa.PORJUGADOR){
            for (Jugador jugador: jugadores){
                if (jugador != jugadorActual){
                    jugador.modificarSaldo(cartaActual.getValor());
                }else
                    jugadorActual.modificarSaldo(cartaActual.getValor());
            }
        }
        
        if (cartaActual.getTipo() == TipoSorpresa.SALIRCARCEL){
            jugadorActual.setCartaLibertad(cartaActual);
        }else{
            // Ver después porque no lo tengo claro
            mazo.add(cartaActual);
        }
        return tienePropietario;
    }
    
    public boolean cancelarHipoteca(Casilla casilla){
        boolean cancelarHipoteca = false;
        if (casilla.estaHipotecada() && jugadorActual.puedoPagarHipoteca(casilla)){
            int cantidadRecibida = casilla.getCosteHipoteca();
            jugadorActual.modificarSaldo(-cantidadRecibida); 
            cancelarHipoteca = true;
        }
        return cancelarHipoteca; 
    }
    
    public boolean comprarTituloPropiedad(){
        boolean puedoComprar = jugadorActual.comprarTitulo();
        return puedoComprar;
    }
    
    public boolean edificarCasa(Casilla casilla){
        boolean puedoEdificar = false;
        if (casilla.soyEdificable()){
            boolean sePuedeEdificar = casilla.sePuedeEdificarCasa();
            if (sePuedeEdificar){
                puedoEdificar = jugadorActual.puedoEdificarCasa(casilla);
                if(puedoEdificar){
                    int costeEdificarCasa = casilla.edificarCasa();
                    jugadorActual.modificarSaldo(-costeEdificarCasa);
                }
            }
        }
        return puedoEdificar;
    }
    
    public boolean edificarHotel(Casilla casilla){
        boolean puedoEdificar = false;
        if (casilla.soyEdificable()){
            boolean sePuedeEdificar = casilla.sePuedeEdificarHotel();
            if (sePuedeEdificar){
                puedoEdificar = jugadorActual.puedoEdificarHotel(casilla);
                if(puedoEdificar){
                    int costeEdificarHotel = casilla.edificarHotel();
                    jugadorActual.modificarSaldo(-costeEdificarHotel);
                }
            }
        }
        return puedoEdificar;
    }
    
    public Sorpresa getCartaActual(){
        return cartaActual;
    }
    
    public Jugador getJugadorActual(){
        return jugadorActual;
    }
    
    public ArrayList<Jugador> getJugadores(){
        return jugadores;
    }
    
    public boolean hipotecarPropiedad(Casilla casilla){
        boolean puedoHipotecar = false;
        if (casilla.soyEdificable()){
            boolean sePuedeHipotecar = !casilla.estaHipotecada();
            if (sePuedeHipotecar){
                puedoHipotecar = jugadorActual.puedoHipotecar(casilla);
                if (puedoHipotecar){
                    int cantidadRecibida = casilla.hipotecar();
                    jugadorActual.modificarSaldo(cantidadRecibida);
                }
            }
        }
        return puedoHipotecar;
    }
    
    public void inicializarJuego(ArrayList<String> nombres){
        inicializarJugadores(nombres);
        inicializarTablero();
        inicializarCartasSorpresa();
        salidaJugadores();
    }
    
    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo){
        boolean libre = false;
        if (metodo == MetodoSalirCarcel.TIRANDODADO){
            int valorDado = dado.tirar();
            libre = valorDado>5;
        }else{
            boolean tengoSaldo = jugadorActual.pagarLibertad(Qytetet.PRECIO_LIBERTAD);
            libre = tengoSaldo;
        }
        if (libre)
            jugadorActual.setEncarcelado(false);
        return libre;
    }
    
    public boolean jugar(){
        int valorDado = dado.tirar();
        Casilla casillaPosicion = jugadorActual.getCasillaActual();
        Casilla nuevaCasilla = tablero.obtenerNuevaCasilla(casillaPosicion,valorDado);
        boolean tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
        if (!nuevaCasilla.soyEdificable()){
            if (nuevaCasilla.getTipo() == TipoCasilla.JUEZ){
                encarcelarJugador();
            }else if (nuevaCasilla.getTipo() == TipoCasilla.SORPRESA){
                cartaActual = mazo.get(0);
                // Se borra la carta del mazo
                mazo.remove(cartaActual);
            }
        }
        return tienePropietario;
    }
    
    public Map<String, Integer> obtenerRanking(){
        Map<String, Integer> ranking = new LinkedHashMap<String, Integer>();
        for (Jugador jugador: jugadores){
            int capital = jugador.obtenerCapital();
            // Habría que añadir el nombre y el capital, pero ¿cómo?
            ranking.put(jugador.getNombre(), capital);
        }
        return ranking;
    }
    
    //Añadido getCasilla, hecho a partir del de Jugador
    public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas){
        ArrayList<TituloPropiedad> titulos = jugadorActual.obtenerPropiedadesHipotecadas(hipotecadas);
        ArrayList<Casilla> casillas = new ArrayList();
        for (TituloPropiedad t: titulos){
            casillas.add(t.getCasilla());
        }
        return casillas;
    }
    
    public Jugador siguienteJugador(){
        int indice = jugadores.indexOf(jugadorActual);
        indice +=1;
        indice %= jugadores.size();
        jugadorActual = jugadores.get(indice);
        return jugadorActual;
    }
    
    public boolean venderPropiedad(Casilla casilla){
        boolean puedoVender = false;
        if (casilla.soyEdificable()){
            puedoVender = jugadorActual.puedoVenderPropiedad(casilla);
            if (puedoVender){
                jugadorActual.venderPropiedad(casilla);
            }
        }
        return puedoVender;
    }
    
    private void encarcelarJugador(){
        if (!jugadorActual.tengoCartaLibertad()){
            Casilla casillaCarcel = tablero.getCarcel();
            jugadorActual.irACarcel(casillaCarcel);
        }else{
            Sorpresa carta = jugadorActual.devolverCartaLibertad();
            mazo.add(carta);
            // Revisar
        }
    }
    
    private void inicializarCartasSorpresa(){
        mazo.add(new Sorpresa("Han pillado alguno de tus chanchullos. Ve a la cárcel.",TipoSorpresa.IRACASILLA,tablero.getCarcel().getNumeroCasilla()));
        mazo.add(new Sorpresa("Un extraño tornado te arrastra hasta la casilla 13.",TipoSorpresa.IRACASILLA,13));
        mazo.add(new Sorpresa("El poder de la fuerza te atrae hasta la casilla 2.",TipoSorpresa.IRACASILLA,2));
        mazo.add(new Sorpresa("Quedas libre de la cárcel. No es aplicable en la vida real",TipoSorpresa.SALIRCARCEL,0));
        
        mazo.add(new Sorpresa("Al fin han reconocido tu talento y tu belleza... Recibes dinero.",TipoSorpresa.PAGARCOBRAR,200));
        mazo.add(new Sorpresa("Pagas al banco con la esperanza de que no descubran lo que te traes entre manos.",TipoSorpresa.PAGARCOBRAR,-250));
        
        mazo.add(new Sorpresa("Recibes dinero por cada una de tus maravillosas propiedades.",TipoSorpresa.PORCASAHOTEL,150));
        mazo.add(new Sorpresa("No cumples la normativa, así que debes pagar por cada una de tus odiosas propiedades.",TipoSorpresa.PORCASAHOTEL,-175));
        // No estoy segura de si están bien los signos, por el método aplicarSorpresa 
        mazo.add(new Sorpresa("Los demás jugadores sospechan de ti. Los sobornas para que no te delaten.",TipoSorpresa.PORJUGADOR,-100));
        mazo.add(new Sorpresa("El resto de jugadores te indemniza para comprar tu silencio.",TipoSorpresa.PORJUGADOR,100));
        Collections.shuffle(mazo);
    }
    
    private void inicializarJugadores(ArrayList<String> nombres){
        Jugador j;
        for (String n: nombres){
            j = new Jugador(n);
            jugadores.add(j);
        }
    }
    
    private void inicializarTablero(){
        tablero = new Tablero();
    }
    
    // Suponemos saldo a 7500
    private void salidaJugadores(){
        for (Jugador j: jugadores){
            j.setCasillaActual(tablero.obtenerCasillaNumero(0));
        }
        int numero = (int) (Math.random()*jugadores.size());
        jugadorActual = jugadores.get(numero);
    }
    
    @Override
    public String toString(){
        return "Qytetet {" + "carta actual=" + cartaActual + ", cartas sorpresa=" + mazo +
                ", jugador actual=" + jugadorActual + ", jugadores=" + jugadores + ", tablero="+
                tablero + ", dado=" + dado + "}\n";
    
    }
    
}

File: QytetetRuby/lib/qytetet.rb
#encoding: utf-8

module ModeloQytetet
  
  require "singleton"
  
  class Qytetet
    
    include Singleton

    MAXJUGADORES = 4
    MAXCARTAS = 10
    MAXCASILLAS = 20
    PRECIOLIBERTAD = 200
    SALDOSALIDA = 1000
    
    def initialize
      @cartaActual = nil
      @mazo = Array.new
      @jugadores = Array.new
      @jugadorActual = nil
      @tablero = nil
      @dado = Dado.instance
    end
    
    def aplicarSorpresa
      tienePropietario = false
      if (@cartaActual.tipo == TipoSorpresa::PAGARCOBRAR)
        @jugadorActual.modificarSaldo(@cartaActual.valor)
      elsif(@cartaActual.tipo == TipoSorpresa::IRACASILLA)
        esCarcel = @tablero.esCasillaCarcel(@cartaActual.valor)
        if (esCarcel)
          encarcelarJugador
        else
          nuevaCasilla = @tablero.obtenerCasillaNumero(@cartaActual.valor)
          tienePropietario = @jugadorActual.actualizarPosicion(nuevaCasilla)
        end
      elsif (@cartaActual.tipo == TipoSorpresa::PORCASAHOTEL)
        @jugadorActual.pagarCobrarPorCasaYHotel(@cartaActual.valor)
      elsif (@cartaActual.tipo == TipoSorpresa::PORJUGADOR)
        @jugadores.each{ |jugador|
          if (jugador != @jugadorActual)
            jugador.modificarSaldo(@cartaActual.valor)
            @jugadorActual.modificarSaldo(-@cartaActual.valor)
          end
        }
      end
        
      if (@cartaActual.tipo == TipoSorpresa::SALIRCARCEL)
        @jugadorActual.setCartaLibertad(@cartaActual)
      else
        # Ver después porque no lo tengo claro
        @mazo << @cartaActual
      end
      tienePropietario
    end
    
    def cancelarHipoteca(casilla)
      cancelarHipoteca = false
      if (casilla.estaHipotecada && @jugadorActual.puedoPagarHipoteca(casilla))
        cantidadRecibida = casilla.getCosteHipoteca
        @jugadorActual.modificarSaldo(-cantidadRecibida) 
        cancelarHipoteca = true
      end
      cancelarHipoteca 
    end
    
    def comprarTituloPropiedad
      puedoComprar = @jugadorActual.comprarTitulo
      puedoComprar
    end
    
    def edificarCasa(casilla)
      puedoEdificar = false
      if (casilla.soyEdificable)
        sePuedeEdificar = casilla.sePuedeEdificarCasa
        if (sePuedeEdificar)
          puedoEdificar = @jugadorActual.puedoEdificarCasa(casilla)
          if(puedoEdificar)
            costeEdificarCasa = casilla.edificarCasa
            @jugadorActual.modificarSaldo(-costeEdificarCasa)
          end
        end
      end
      puedoEdificar
    end
    
    def edificarHotel(casilla)
      puedoEdificar = false
      if (casilla.soyEdificable)
        sePuedeEdificar = casilla.sePuedeEdificarHotel
        if (sePuedeEdificar)
          puedoEdificar = @jugadorActual.puedoEdificarHotel(casilla)
          if(puedoEdificar)
            costeEdificarHotel = casilla.edificarHotel
            @jugadorActual.modificarSaldo(-costeEdificarHotel)
          end
        end
      end
      puedoEdificar
    end
    
    def getCartaActual
      @cartaActual
    end
    
    def getJugadorActual
      @jugadorActual
    end
    
    def getJugadores
      @jugadores
    end
    
    def hipotecarPropiedad(casilla)
      puedoHipotecar = false
      if (casilla.soyEdificable)
        sePuedeHipotecar = !casilla.estaHipotecada()
        if (sePuedeHipotecar)
          puedoHipotecar = @jugadorActual.puedoHipotecar(casilla)
          if (puedoHipotecar)
            cantidadRecibida = casilla.hipotecar
            @jugadorActual.modificarSaldo(cantidadRecibida)
          end
        end
      end
      puedoHipotecar
    end
    
    def inicializarJuego(nombres)
      inicializarJugadores(nombres)
      inicializarTablero
      inicializarCartasSorpresa
      salidaJugadores
    end
    
    def intentarSalirCarcel(metodo)
      libre = false
      if (metodo == MetodoSalirCarcel::TIRANDODADO)
        valorDado = @dado.tirar
        libre = valorDado>5
      else
        tengoSaldo = @jugadorActual.pagarLibertad(-Qytetet::PRECIO_LIBERTAD)
        libre = tengoSaldo 
      end
      if (libre)
        @jugadorActual.setEncarcelado(false)
      end
      libre
    end
    
    def jugar
      valorDado = @dado.tirar
      casillaPosicion = @jugadorActual.getCasillaActual
      nuevaCasilla = @tablero.obtenerNuevaCasilla(casillaPosicion,valorDado)
      tienePropietario = @jugadorActual.actualizarPosicion(nuevaCasilla)
      if (!nuevaCasilla.soyEdificable)
        if (nuevaCasilla.tipo == TipoCasilla::JUEZ)
          encarcelarJugador
        elsif (nuevaCasilla.tipo == TipoCasilla::SORPRESA)
          @cartaActual = @mazo[0]
          # Se borra la carta del mazo
          @mazo.delete(@cartaActual)
        end
      end
      tienePropietario
    end
    
    def obtenerRanking
      ranking = LinkedHashMap<String, Integer>().new
      @jugadores.each { |jugador|
        capital = jugador.obtenerCapital
        # Habría que añadir el nombre y el capital, pero ¿cómo?
        ranking.put(jugador.getNombre, capital)
      }
      ranking
    end
    
    #Añadido getCasilla, hecho a partir del de Jugador
    def propiedadesHipotecadasJugador(hipotecadas)
      titulos = @jugadorActual.obtenerPropiedadesHipotecadas(hipotecadas)
      casillas = Array.new
      titulos.each{ |titulo|
        casillas << (titulo.getCasilla)
      }
      casillas
    end
    
    def siguienteJugador
      indice = @jugadores.index(@jugadorActual)
      indice +=1
      indice %= @jugadores.size
      @jugadorActual = @jugadores[indice]
      @jugadorActual
    end
    
    def venderPropiedad(casilla)
      puedoVender = false
      if (casilla.soyEdificable)
        puedoVender = @jugadorActual.puedoVenderPropiedad(casilla)
        if (puedoVender)
          @jugadorActual.venderPropiedad(casilla)
        end
      end
      puedoVender
    end
    
    private
    
    def encarcelarJugador
      if (@jugadorActual.tengoCartaLibertad)
        casillaCarcel = @tablero.carcel
        @jugadorActual.irACarcel(casillaCarcel)
      else
        carta = @jugadorActual.devolverCartaLibertad
        @mazo << (carta)
        # Revisar
      end
    end
    
    def inicializarCartasSorpresa
      @mazo << Sorpresa.new("Han pillado alguno de tus chanchullos. Ve a la cárcel.",TipoSorpresa::IRACASILLA,@tablero.carcel.numeroCasilla)
      @mazo << Sorpresa.new("Un extraño tornado te arrastra hasta la casilla 13.",TipoSorpresa::IRACASILLA,13)
      @mazo << Sorpresa.new("El poder de la fuerza te atrae hasta la casilla 2.",TipoSorpresa::IRACASILLA,2)
      @mazo << Sorpresa.new("Quedas libre de la cárcel. No es aplicable en la vida real",TipoSorpresa::SALIRCARCEL,0)
        
      @mazo << Sorpresa.new("Al fin han reconocido tu talento y tu belleza... Recibes dinero.",TipoSorpresa::PAGARCOBRAR,200)
      @mazo << Sorpresa.new("Pagas al banco con la esperanza de que no descubran lo que te traes entre manos.",TipoSorpresa::PAGARCOBRAR,-250)
        
      @mazo << Sorpresa.new("Recibes dinero por cada una de tus maravillosas propiedades.",TipoSorpresa::PORCASAHOTEL,150)
      @mazo << Sorpresa.new("No cumples la normativa, así que debes pagar por cada una de tus odiosas propiedades.",TipoSorpresa::PORCASAHOTEL,-175)
        
      @mazo << Sorpresa.new("Los demás jugadores sospechan de ti. Los sobornas para que no te delaten.",TipoSorpresa::PORJUGADOR,-100)
      @mazo << Sorpresa.new("El resto de jugadores te indemniza para comprar tu silencio.",TipoSorpresa::PORJUGADOR,100)
      @mazo.shuffle!
    end
    
    def inicializarJugadores(nombres)
      nombres.each{ |j|
            n = Jugador.new(j)
            @jugadores << n
      } 
    end
    
    def inicializarTablero
      @tablero = Tablero.new
    end
    
    def salidaJugadores
      @jugadores.each{ |j|
        j.setCasillaActual(@tablero.obtenerCasillaNumero(0))
      }
      # Comprobar
      numero = rand(@jugadores.size)
      @jugadorActual = @jugadores[numero]
    end
    
    public
    
    def to_s
        "Qytetet \n carta actual= #{@cartaActual} \n cartas sorpresa= #{@mazo} \n
                jugador actual= #{@jugadorActual} \n jugadores= #{@jugadores} \n tablero=
                #{@tablero} \n dado= #{@dado} \n"
    end
    
  end
end
